context/README.md

```diff
@@ -0,0 +1,108 @@
+# Implicit Context Propagation
+
+Much has been said about context propagation for the gRPC Python asyncio API.
+This is an example demonstrating my proposal for the shape of an API supporting
+implicit context propagation.
+
+The core of the API is really two functions (whose names I leave open to
+bikeshedding):
+
+```python
+# module grpc.aio
+def get_context_value(key: Text) -> Optional[Text]:
+    """Get a value associated with a key from the coroutine-local context."""
+
+def set_context_value(key: Text, value: Text):
+    """Sets a value associated with a key in the coroutine-local context."""
+```
```

Comment:
`contextvars` already provides get and set methods [Python Doc](https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.get). Can you explain more about the benefits of including such mechanism in our framework?

Also, I thought it is up to the tracing library to choose propagation mechanism, since the user does not need to know much about it? What if the tracing library sets context variable at the start of an RPC, and then provides metric recording helper to collect metrics, and finally before the finish of RPC, dump the data to metadata?

Reply:
This makes the use of `contextvars` an implementation detail. So, while we gain the useful semantics of the module, we are not married to any unintended aspects of their API.

> What if the tracing library sets context variable at the start of an RPC, and then provides metric recording helper to collect metrics, and finally before the finish of RPC, dump the data to metadata?

That use case is supported by this API. They would use `set_context_value()` in the server interceptor. Then, as they want to add data while the RPC is being processed, they can add to that value by doing:
```
old_value = get_context_value('foo')
new_value = data_added(old_value)
set_context_value('foo', new_value)
```
Finally, in the client interceptor, they use `get_context_value` to retrieve the value to add to the metadata.

You could make the argument that returning `Text` might not be the best interface for mutation or storage of structured data. I'd agree on that point. But that would just make me lean toward an interface with `Any` instead of `Text`.

At any rate, I don't think mutating a value in the context will be a common use case.

Reply:
Is it possible to let library authors to use `contextvars` directly without we introducing helper functions in gRPC, and achieve the same functionality?